hudi-spark-datasource/hudi-spark3.2plus-common/src/main/scala/org/apache/spark/sql/hudi/catalog/HoodieCatalog.scala

```diff
@@ -369,4 +378,28 @@ object HoodieCatalog {
 
     (identityCols, bucketSpec)
   }
+
+  def buildPartitionTransforms(spark: SparkSession,
+                               basePath: String): Array[Transform] = {
+    val metaClient = HoodieTableMetaClient.builder()
+      .setConf(spark.sessionState.newHadoopConf())
+      .setBasePath(basePath)
+      .build()
```

Comment:
Do you know under what conditions the stage change methods are called?  This method has non-trivial latency.  If often called, this may not scale.

Reply:
From debugging the tests, I see that this is called just once when the table is loaded. I think this should be fine. Even in other query engines, we build metaclient once per query.